Dmabuf IPC support for amdgpu plugin (v2)

fdavid-amd opened the pull request:
This patch set, in combination with the accompanying kernel patch set, is the
second draft of work to support amdgpu dmabuf IPC. Since the first post, the
patches have been split to make them more readable and made
backwards-compatible.

fdavid-amd committed:
restorer: Skip non-regular VMAs

amdgpu represents allocated device memory as a memory mapping
of the device file. This is a non-standard VMA that must
be handled by the plugin, not the normal VMA code.

Ignore all VMAs on device files.

Signed-off-by: David Francis <David.Francis@amd.com>

fdavid-amd committed:
plugin/amdgpu: Add handling for amdgpu drm buffer objects

Buffer objects held by the amdgpu drm driver are checkpointed with
the new BO_INFO and MAPPING_INFO ioctls/ioctl options. Handling
is in amdgpu_plugin_drm.h

Handling of imported buffer objects may require dmabuf fds to be
transferred between processes. These occur over fdstore, with the
handle-fstore id relationships kept in shread memory. There is a
new plugin callback: RESTORE_INIT to create the shared memory.

During checkpoint, track shared buffer objects, so that buffer objects
that are shared across processes can be identified.

During restore, track which buffer objects have been restored. Retry
restore of a drm file if a buffer object is imported and the
original has not been exported yet. Skip buffer objects that have
already been completed or cannot be completed in the current restore.

So drm code can use sdma_copy_bo, that function no longer requires
kfd bo structs

Update the protobuf messages with new amdgpu drm information.

Signed-off-by: David Francis <David.Francis@amd.com>

fdavid-amd committed:
files-ext: Allow plugin files to retry

amdgpu dmabuf CRIU requires the ability of the amdgpu plugin
to retry.

Change files_ext.c to read a response of 1 from a plugin restore
function to mean retry.

Signed-off-by: David Francis <David.Francis@amd.com>

fdavid-amd committed:
plugin/amdgpu: Add amdgpu drm header

For amdgpu plugin to call the new amdgpu drm CRIU ioctls, it needs
the amdgpu drm header file, copied from the kernel's includes.

Signed-off-by: David Francis <David.Francis@amd.com>

avagin commented on criu/files.c:
```diff
@@ -836,6 +836,9 @@ struct fdinfo_list_entry *collect_fd_to(int pid, FdinfoEntry *e, struct rst_info
 	struct fdinfo_list_entry *new_le;
 
+	if (fdesc->ops->type == FD_TYPES__EXT)
+		run_plugins(COLLECT_FILE, pid, fdesc->id);
```
need to check a return code

fdavid-amd replied:
Fixed

avagin commented on criu/cr-dump.c:
```diff
@@ -2225,6 +2225,9 @@ int cr_dump_tasks(pid_t pid)
 			goto err;
 	}
 
+	if(run_plugins(DUMP_DEVICE_LATE, pid))
```
run_plugins() returns -ENOTSUP if no hook is registered. This is why ci tests fail on this pr.

fdavid-amd replied:
Fixed

avagin commented on criu/servicefd.c:
```diff
@@ -312,5 +313,15 @@ int clone_service_fd(struct pstree_item *me)
 	service_fd_id = id;
 	ret = 0;
 
+	next_high_fd = service_fd_base + 1024;
```
Where does this magic number 1024 come from? Perhaps I've missed something, but it doesn't seem safe to assume that next_high_fd will be available and won't conflict with other descriptors. I feel like a detailed comment is needed here.

avagin commented on criu/servicefd.c:
```diff
@@ -312,5 +313,15 @@ int clone_service_fd(struct pstree_item *me)
+{
+	if (next_high_fd > service_fd_rlim_cur)
+		return -1;
+	next_high_fd += 1;
```
Why don't you use the fdstore for temporary file descriptors? It was created specifically for this purpose.

fdavid-amd replied:
Thanks, didn't know about that. It seems like it would work for the dmabuf_fds, but I would still need to put the socket fd itself somewhere.

fdavid-amd committed:
plugin/amdgpu: Add drm header

The amdgpu plugin usually calls drm ioctls through the libdrm
wrappers. However, amdgpu restore requires dealing with dmabufs
and gem handles directly, which means drm ioctls must be
called directly.

Add the drm.h header (from the kernel's uapi).

Signed-off-by: David Francis <David.Francis@amd.com>

fdavid-amd committed:
plugin: Add DUMP_DEVICES_LATE callback

The amdgpu plugin was counting how many files were checkpointed
to determine when it should close the device files.

The number of device files is not consistent; a process may
have multiple copies of the drm device files open.

Instead of doing this counting, add a new callback after all
files are checkpointed, so plugins can clean up their
resources at an appropriate time.

Signed-off-by: David Francis <David.Francis@amd.com>

avagin commented:
@fdavid-amd how is it going? Let me know if you need help with this pr.

fdavid-amd commented:
@avagin 
Sorry for leaving this untouched for so long, I was revising the kernel side of the changes. Could you take a look at this again?

fdavid-amd commented on plugins/amdgpu/criu-amdgpu.proto:
```diff
@@ -61,6 +62,29 @@ message criu_kfd {
+
 message criu_render_node {
 	required uint32 gpu_id = 1;
+	required uint32 id = 2;
```
That is true - I wasn't aware that was a type of backwards-compatibility we committed to.

fdavid-amd commented on plugins/amdgpu/amdgpu_plugin_util.c:
```diff
@@ -49,37 +51,149 @@ struct tp_system dest_topology;
 
-	/* We add ONE to include checkpointing of KFD device */
-	dev_file_cnt = 1 + topology_gpu_count(topo);
+	if (handle == -1) {
```
Fixed

fdavid-amd commented on plugins/amdgpu/amdgpu_plugin_util.c:
```diff
@@ -49,37 +51,149 @@ struct tp_system dest_topology;
+       if (len_read != buf_len) {
+               pr_err("Unable to read file (read:%ld buf_len:%ld)\n", len_read, buf_len);
+               return -EIO;
+
```
Fixed

fdavid-amd commented:
Some of the build failures I don't understand - they might be an issue with this patch set or they might be an issue with the CI. There's a lot of 'Error: ipv6: address already assigned.', which could possibly be a result of the way that I'm allocating the plugin socket but seems unlikely. 
The linter errors I do understand - they're flagging real grammar and syntax errors in the amdgpu-drm.h and drm.h files - but the point of those files is that they're exact copies of the kernel header files of the same names, so I don't want to change them.

fdavid-amd commented on criu/include/criu-plugin.h:
```diff
@@ -64,6 +64,8 @@ enum {
 
 	CR_PLUGIN_HOOK__COLLECT_FILE = 13,
 
+	CR_PLUGIN_HOOK__DUMP_DEVICE_LATE = 14,
```
Fixed

fdavid-amd commented:
@avagin sorry for the repeat ping, but there is something of a rush on my part to get kernel patch https://lists.freedesktop.org/archives/dri-devel/2025-July/514922.html merged in the next week or so (it adds the CHANGE_HANDLE ioctl). As part of merging that patch, I need to demonstrate that it has written and reasonably stable userspace code that uses it (which would be these CRIU patches). It would be a big help to me if you could say if the general approach taken by these patches is acceptable or unacceptable. 
Thanks,
David

avagin commented:
@fdavid-amd The general approach looks good to me. We just need to adjust how we manage file descriptors, but it is just a criu internal thing.

avagin commented:
> It would be a big help to me if you could say if the general approach taken by these patches is acceptable or unacceptable.

BTW we can merge the current version to the criu-dev branch and do all other adjustments on top of it.

fdavid-amd commented:
This code is currently not functional because it relies on ioctls that are not merged into any kernel branch; is is still okay to merge?

rst0git commented:
>This code is currently not functional because it relies on ioctls that are not merged into any kernel branch; is is still okay to merge?

The way we usually handle this problem in CRIU is by checking if the required kernel functionality is available and disabling features that are not supported. However, it might be better to wait until the ioctls are merged.

fdavid-amd commented:
Some of these compilation tests seem to be failing at a install phase; is that something my code is doing or is it a problem with the test?

fdavid-amd commented:
@avagin There's now an additional patch in this patch set (Support for Open dmabuf handles) to handle the case where a dmabuf fd is open during checkpoint. It involves some changes to core CRIU, what do you think of the approach?

fdavid-amd commented on criu/files.c:
```diff
@@ -329,7 +329,9 @@ uint32_t make_gen_id(uint32_t st_dev, uint32_t st_ino, uint64_t pos)
-int do_dump_gen_file(struct fd_parms *p, int lfd, const struct fdtype_ops *ops, FdinfoEntry *e)
+/* Use "force" to override cache */
+int do_dump_gen_file(struct fd_parms *p, int lfd,
+					const struct fdtype_ops *ops, FdinfoEntry *e, bool force)
```
CRIU has always dealt with dependencies between restoring files, but dmabuf fd adds dependencies between dumping files. A dmabuf fd can't be dumped until we've dumped the device file it was exported from. For this reason this patch set adds a retry system to file dumping. Files that are being retried should have the appropriate dump call made even if there is no new file descriptor for them.

I have implemented tests for internal testing, which can be found here. The DmabufDelayTest in this branch pauses while a dmabuf fd is active to trigger this code.
https://github.com/fdavid-amd/ROCT-Thunk-Interface-CRIU#

avagin commented on criu/files.c:
```diff
@@ -654,9 +661,32 @@ int dump_task_files_seized(struct parasite_ctl *ctl, struct pstree_item *item, s
-			ret = dump_one_file(item->pid, dfds->fds[i + off], lfds[i], opts + i, ctl, &e, dfds);
+			ret = dump_one_file(item->pid, dfds->fds[i + off], lfds[i], opts + i, ctl, &e, dfds, false);
+			if (ret == -EAGAIN) {
+				retry_indices[retry_count++] = i;
```
i+off?

fdavid-amd replied:
I believe this addition is performed a few lines later

fdavid-amd commented on criu/files.c:
```diff
@@ -654,9 +661,32 @@ int dump_task_files_seized(struct parasite_ctl *ctl, struct pstree_item *item, s
 			if (ret)
 				break;
+		}
+		/* Dmabuf FDs cannot be dumped unless we have first dumped the device file that exported them.
```
I was under the impression that CRIU didn't close the files it was dumping with until later in the process. 
The earlier patches in this pr allow the amdgpu plugin to find the dmabuf fd even if it was exported by another process.

fdavid-amd committed:
plugin/amdgpu: Support for checkpoint of dmabuf fds

amdgpu libraries that use dmabuf fd to share GPU memory between
processes close the dmabuf fds immediately after using them.
However, it is possible that checkpoint of a process catches one
of the dmabuf fds open. In that case, the amdgpu plugin needs
to handle it.

The checkpoint of the dmabuf fd does require the device file
it was exported from to have already been dumped

To identify which device this dmabuf fd was exprted from, attempt
to import it on each device, then record the dmabuf handle
it imports as. This handle can be used to restore it.

Signed-off-by: David Francis <David.Francis@amd.com>

avagin commented:
> To handle this, the patch "Allow for retry of dumped files" racks files that failed to dump and tries to dump them again after all other dumps have succeeded. It uses a variable "is_retry" so that dump will be re-attempted even if it failed the first time.

Why do we need to do that in the criu generic code? Why can't we cache these file descriptors in the "plugin" and then dump them when their device file descriptor is dumped?

avagin commented:
LGTM

rst0git commented:
@fdavid-amd Would it be possible to provide more information on how we can test these patches?

rst0git commented on criu/include/criu-plugin.h:
```diff
@@ -62,6 +62,8 @@ enum {
 
 	CR_PLUGIN_HOOK__POST_FORKING = 12,
 
+	CR_PLUGIN_HOOK__RESTORE_INIT = 13,
```
It would be good to update [plugins/amdgpu/README.md](https://github.com/checkpoint-restore/criu/blob/a000d89c70611e83dcc113ae59d0e4fdd244237d/plugins/amdgpu/README.md) with information about the new plugin callbacks.

rst0git commented on plugins/amdgpu/amdgpu_plugin_dmabuf.c:
```diff
@@ -0,0 +1,201 @@
+       pr_err("Not all dma buffers have been dumped\n");
+       return -1;
+   }
+   return 1;
```
Why do we use `return 1;` here?

fdavid-amd commented on plugins/amdgpu/amdgpu_plugin.c:
```diff
@@ -1826,12 +1839,17 @@ int amdgpu_plugin_restore_file(int id, bool *retry_needed)
+		if (!img_fp) {
+			ret = amdgpu_plugin_dmabuf_restore(id);
+			if (ret == 1) {
+				*retry_needed = true;
```
Yes, done

rst0git commented:
@fdavid-amd I've applied the patches here with @avagin's changes from the `wip/amdgpu` branch in the following pull request: https://github.com/checkpoint-restore/criu/pull/2810
